src/api/dashboard.js

```diff
@@ -56,14 +63,29 @@ Dashboard.prototype = {
     this._dashboard.vis.mapvis.map.setBounds([state.map.ne, state.map.sw]);
   },
 
-  onStateChanged: function (callback, shareURLs) {
-    this._dashboard.vis.once('dataviewsFetched', function () {
-      this._dashboard.widgets._widgetsCollection.initialState();
-      this._dashboard.widgets._widgetsCollection.each(function (m) {
-        m.applyInitialState();
-      }, this);
-      shareURLs === true && this._bindChange(callback);
-    }, this);
+  _widgetsAlreadyInitialized: function () {
+    var widgetsCollection = this._dashboard.widgets._widgetsCollection;
+    if (widgetsCollection.size() > 0) {
+      return this._dashboard.widgets._widgetsCollection.hasInitialState();
+    }
+    return true;
+  },
+
+  onDataviewsFetched: function (callback) {
+    var areDataviewsFetched = this._widgetsAlreadyInitialized();
+    if (areDataviewsFetched) {
+      callback && callback();
+    } else {
+      this._dashboard.vis.once('dataviewsFetched', function () {
```

Comment:
We could even move this method to CartoDB.js since it's basically asking about dataviews (a concept that belongs to CartoDB.js):

```
onStateChanged: function (callback) {
  if (this._dashboard.vis.onDataviewsFetched(callback)) {
    // something....
  });
},
```